### PR TITLE
Provide JS smoke tests for the WAYF

### PR DESCRIPTION
**Pull request by MKodde**
This PR will add JS smoke tests for the WAYF

It was hard to test the features using the Behat tests. The introduction of Puppeteer and Jest makes is easier to test for regressions.

In order to test the WAYF, a test endpoint was added. This endpoint will be used in a headless browser on travis and Puppeteer & Jest will run tests against it.

### CODE REVIEW

**Comment by MKodde** [comment]
Where do we stand at this point? 

1. The WAYF endpoint is now available in test mode. 
1. Instructions where added to document the configuration options for the test WAYF endpoint.
1. A very minimal POC is provided that unit tests can be added to the JS that can be found in the JS folder.
1. A more complete smoke test was added to the project testing the functionality of a 'basic' WAYF.

Todo in next PR: provide more smoke test coverage for different WAYF configuration options (disconnected, request access, filtering, cutoff, and so forth)

**Comment by MKodde** [comment]
After my previous update some changes have been applied:
1. The test coverage has been increased, the different WAYF options for a WAYF with only connected IdPs has been covered in full. 
2. The expect-puppeteer lib has been added to the mix in order to make assertion expressions more intuitive. And reap the benefit of its built-in 'wait for the element to actually be on the page' feature.

In the next PR, the test coverage will be made more substantial. What will be covered there:
1. Test the correct working of the keyboard navigation
2. Test unconnected IdP features like 'request access'
3. See if we can incorporate more mouse interaction in order to verify the DOM is rendered correctly.

File: src/OpenConext/EngineBlockFunctionalTestingBundle/Controllers/WayfController.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        $rememberChoiceFeature = (bool) $request->get('rememberChoiceFeature', false);
+        $cutoffPointForShowingUnfilteredIdps = $request->get('cutoffPointForShowingUnfilteredIdps', 100);
+
+        $connectedIdps = (int) $request->get('connectedIdps', 5);
+        $unconnectedIdps = (int) $request->get('unconnectedIdps', 0);
+
+        return new Response($this->twig->render(
+            '@theme/Authentication/View/Proxy/wayf.html.twig',
+            [
+                'action' => 'https://continue-to-idp.example.com',
```
Comment (pablothedude) [inline src/OpenConext/EngineBlockFunctionalTestingBundle/Controllers/WayfController.php]: Couldn't  we use the ways endpoint itself to test the outcome? I think it should be very valuable to know if this action is ultimately done.

File: theme/material/javascripts/tests/smoke/wayf/WayfShowsConnectedIdps.test.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    'WAYF is showing connected IdPs by default',
+    () => {
+        let page;
+        beforeAll(async () => {
+            page = await global.__BROWSER__.newPage();
+            const override = Object.assign(page.viewport(), {width: 1400, height: 1000});
+            page.setViewport(override);
+        }, timeout);
+
+        it('Should show five connected IdPs and the search field', async () => {
```
Comment (pablothedude) [inline theme/material/javascripts/tests/smoke/wayf/WayfShowsConnectedIdps.test.js]: A very valuable initial test, maybe we could verify with our PO what he would like to test additionally.

**Comment by pablothedude** [comment]
A nice POC which will be very valuable! Great work!

I think puppeteer will do the job just fine, but shouldn't it be more valuable if we could also support other browsers then just chrome? I think that should be more valuable because then we could test a lot of browsers at once and have those use cases covered as well.

File: .travis.yml [hunk 3]
```diff
@@ -61,6 +61,9 @@ before_script:
   - composer config secure-http false
   - composer install --prefer-dist -n -o
   - app/console cache:clear --env=test
+  - cd theme
+  - npm install
+  - cd ..
```
[the remark this answers is not in the document]
Comment (MKodde) [reply .travis.yml]: Good point

File: src/OpenConext/EngineBlockFunctionalTestingBundle/Controllers/WayfController.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        $this->twig = $twig;
+    }
+
+    public function wayfAction(Request $request)
+    {
+        $currentLocale = $request->get('lang', 'en');
+        $request->cookies->set('lang', $currentLocale);
+        $backLink = (bool) $request->get('backLink', false);
+        $displayUnconnectedIdpsWayf = (bool) $request->get('displayUnconnectedIdpsWayf', false);
```
[the remark this answers is not in the document]
Comment (MKodde) [reply src/OpenConext/EngineBlockFunctionalTestingBundle/Controllers/WayfController.php]: Yes, this is possible, I do not see the benefit in doing this.

File: theme/material/javascripts/tests/smoke/wayf/WayfShowsConnectedIdps.test.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+const timeout = 5000;
+
+describe(
+    'WAYF is showing connected IdPs by default',
+    () => {
+        let page;
+        beforeAll(async () => {
+            page = await global.__BROWSER__.newPage();
+            const override = Object.assign(page.viewport(), {width: 1400, height: 1000});
```
[the remark this answers is not in the document]
Comment (MKodde) [reply theme/material/javascripts/tests/smoke/wayf/WayfShowsConnectedIdps.test.js]: Yes, switching to Full HD 1920x1080

File: theme/jest.config.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
```
[the remark this answers is not in the document]
Comment (MKodde) [reply theme/jest.config.js]: A good suggestion, but in my opinion not one for this PR. This one is to loaded as is.

File: src/OpenConext/EngineBlockFunctionalTestingBundle/Controllers/WayfController.php (hunk 1 above)
Comment (MKodde) [reply src/OpenConext/EngineBlockFunctionalTestingBundle/Controllers/WayfController.php]: I don't see this working in our setup? Or do you have an idea of how we could achieve this?

File: theme/material/javascripts/tests/smoke/wayf/WayfShowsConnectedIdps.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (MKodde) [reply theme/material/javascripts/tests/smoke/wayf/WayfShowsConnectedIdps.test.js]: Updated to use full HD viewport size.

File: theme/jest.config.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (MKodde) [reply theme/jest.config.js]: We just discussed this IRL, we both agree that moving to TS is very valuable, but out of scope for this ticket. We should not wait much longer tho. Lets aim for a TS upgrade next time we are touching the WAYF JS code.

**Comment by MKodde** [comment]
We are aiming to include these changes in EB 5.9. But also want to refrain from investing much more effort in this PR. In order to acept this PR, we need the tests to pass. 

`expect-puppeteer` seems to be the cause of the build problems. My aim is to remove said package for now and update the currently used assertions to use the regular Jest/Puppeteer expressions.

**Pull request merged**